godot-fmt/src/lib.rs

```diff
@@ -0,0 +1,350 @@
+///   - colons if they follow braces `..},`
+/// - whitespace around punctuation is tried to be made as natural as possible
+///   - `::` always stick to whatever comes after them `use ::crate::module::{..}`
+///   - `$` is always prefix
+///   - `!` tries to join with whatever comes after it, `!boolean`, `x != y`
+///   - `&` tries to join with whatever comes after it, `&self`, `&'static`, `a && b`
+///   - joint punctuation tries to stay joint
+///   - if an identifier is followed by `.` or `()` then it's "sticky", `f()`, `s.x`
+///
+/// Current "flaws" (still compiles properly, just a bit ugly):
+/// - Because of stickiness, bitwise AND doesn't really look nice, like `a &mask`.
+///   Unless we check for all possible keywords (which would be costly), we
+///   can't solve this.
+/// - Also because of stickiness, macros are a bit weird, `name !`. If we made
+///   all `!` stick to previous identifiers then we'd get `x!= y`.
+///   Maybe that's a good trade-off if your code contains more not-equals checks
+///   than macro invocations.
+/// - anonymous function literals are a bit weird because the `|` don't have
+///   joint spacing when they come out of a `TokenStream` and identifiers don't
+///   have any spacing information attached, so it looks like `| arg |`.
+///   Also to know whether it's a chained bitwise or a closure would need more
+///   information. `move | arg | x`
+/// - Because we don't keep track of what's a type and what's an expression,
+///   generic parameters look a bit spaced out. `T < A, B >`
+/// - `match` arms can be `<pat> => <expr>,` and there _should_ be a line break
+///   after the `,` but because this only looks at one TokenKind at a time
+///   and only keeps a single state, it's too much effort to keep track of which
+///   `,` in a [`TokenGroup`] should trigger a newline (it would involve a stack
+///   of infos whether that group is in a match or not, detecting pattern etc)
+/// - Because `,` can be used for function args (no-newline) and separators
+///   in `struct` and `enum` definitons, those definitions are more awkward
+///   and crammed into one line as a trade-off to make function calls and args
+///   smoother.
+///   Maybe it's worth to make ALL args on their own line, then function calls
+///   get big, but `struct`s and `enum`s would be more natural?
+pub fn format_tokens(tokens: TokenStream) -> String {
+    let mut out = String::new();
+
+    format(FormatState::Start, 0, tokens.into_iter(), &mut out);
+
+    out
+}
+
+fn indent(n: usize) -> &'static str {
+    // This looks strange but it means we don't need to actually allocate anything.
+    // The downside is there's a limit to how deep we can nest.
+    // The code that's generated doesn't seem like it's any deeper than this.
+
+    //                 |   |   |   |   |   |   |   |   |   |   |   |   |   |   |
+    let idents = "                                                        ";
+
+    let end = n * 4;
+    &idents[0..end]
```

Comment:
Nice trick! 🙂

Reply:
Thanks 😅 Now with better formatting that doesn't depend on the type-inlay in vscode to properly align 😬